base/libdl.jl

```diff
@@ -314,4 +317,98 @@ function dllist()
     return dynamic_libraries
 end
 
+
+const LazyLibraryLock = Base.ReentrantLock()
+"""
+    LazyLibrary(name, flags = <default dlopen flags>,
+                dependencies = LazyLibrary[], on_load_callback = nothing)
+
+Represents a lazily-loaded library that opens itself and its dependencies on first usage
+in a `dlopen()`, `dlsym()`, or `ccall()` usage.  While this structure contains the
+ability to run arbitrary code on first load via `on_load_callback`, we caution that this
+should be used sparingly, as it is not expected that `ccall()` should result in large
+amounts of Julia code being run.  You may call `ccall()` from within the
+`on_load_callback` but only for the current library and its dependencies.
+"""
+mutable struct LazyLibrary
+    # Name and flags to open with
+    const name::AbstractString
+    const flags::UInt32
+
+    # Dependencies that must be loaded before we can load
+    dependencies::Vector{LazyLibrary}
+
+    # Function that gets called once upon initial load with the pointer as an argument
+    on_load_callback::Union{Nothing,Function}
```

Comment:
```suggestion
    on_load_callback
```
(likely much faster)